automata/fa/fa.py

```diff
@@ -2,11 +2,245 @@
 """Classes and methods for working with all finite automata."""
 
 import abc
+import itertools
+import os
+import pathlib
+import typing
+import uuid
+from collections import defaultdict
+
+import graphviz
+from coloraide import Color
 
 from automata.base.automaton import Automaton
 
 
 class FA(Automaton, metaclass=abc.ABCMeta):
     """An abstract base class for finite automata."""
 
-    pass
+    @staticmethod
+    def get_state_name(state_data):
+        """
+        Get an string representation of a state. This is used for displaying and
+        uses `str` for any unsupported python data types.
+        """
+        if isinstance(state_data, str):
+            if state_data == "":
+                return "λ"
+
+            return state_data
+
+        if isinstance(state_data, (set, frozenset, list, tuple)):
+            inner = ", ".join(FA.get_state_name(sub_data) for sub_data in state_data)
+            if isinstance(state_data, (set, frozenset)):
+                return '{' + inner + '}'
+
+            if isinstance(state_data, tuple):
+                return '(' + inner + ')'
+
+            if isinstance(state_data, list):
```

Comment:
I don't think this case is needed, lists aren't hashable and can't be put in a Python set.

Reply:
Agreed with @eliotwrobson on this.

Reply:
Technically, the `set` case a few lines above is also unnecessary, but it's better to support them in case someday the hashable-ity requirement of fa-states changes someday or when this visualization is used for other automata. I will do a performance benchmark on all of the code before merging. If there are any notable hits, I will remove this.